Ability to drag(extrude) an edge

sivako opened the pull request:
Drag from any where between 2 points in edge.
![nebula-extrude](https://user-images.githubusercontent.com/1399914/48431142-fd904000-e796-11e8-9c6f-96b78af10b3f.gif)

supersonicclay commented:
Looks pretty good. I think we can make this better by not adding additional points. See this demonstration and compare to how JOSM extrude works:
![extrude-bug](https://user-images.githubusercontent.com/469582/48502868-d4ef6100-e7fd-11e8-8e0b-056e4ca00a62.gif)

supersonicclay commented:
Also, it would be good i think if clicking an edge adds a point. With JOSM, you double-click to add an intermediate point. But I think it is ok to make it a single click, and should be easier to implement.

sivako commented:
Thank you @supersonicclay.  great points. I will update accordingly (and adding point). Thanks again.

sivako commented:
![nebula-extrude-2](https://user-images.githubusercontent.com/1399914/48702491-0eb5d280-ec17-11e8-8917-e453b2fbbfeb.gif)

@supersonicclay, I have updated code addressing comments.  1) while dragging do not add point when 90 corners 2) ability to add point by clicking on edge. (both add and remove point by click will be available for both modify and extrude modes, I think that will be useful for both modes).

sivako commented on examples/deck/example.js:
```diff
@@ -329,7 +330,9 @@ export default class Example extends Component<
         let updatedSelectedFeatureIndexes = this.state.selectedFeatureIndexes;
-        if (!['movePosition', 'rotating', 'translating', 'scaling'].includes(editType)) {
+        if (
+          !['movePosition', 'moveEdge', 'rotating', 'translating', 'scaling'].includes(editType)
```
ok, Thats good idea. I updated to 'extruding' here.

sivako commented on modules/core/src/lib/mode-handlers/extrude-handler.js:
```diff
@@ -0,0 +1,234 @@
+import { getPickedEditHandle } from './mode-handler.js';
+import { ModifyHandler } from './modify-handler';
+
+export class ExtrudeHandler extends ModifyHandler {
```
Added tests for ExtrudeHandler and utils.

sivako commented on modules/core/src/lib/utils.js:
```diff
@@ -51,3 +57,34 @@ export function recursivelyTraverseNestedArrays(
+    coordinates: [p1, p2]
+  };
+  const pt = point(groundCoords);
+  const options = { units: 'miles' };
```
Oh right, No need for options. I have removed.

sivako commented on modules/core/src/lib/mode-handlers/extrude-handler.js:
```diff
@@ -0,0 +1,234 @@
+
+        editAction = {
+          updatedData: updatedData.getObject(),
+          editType: 'addEdge',
```
agree, that make sense. have updated to 'startExtruding', 'extruding' and 'extruded' for start, move and stop events. Updated the doc.

sivako commented:
@supersonicclay, Fixed the issue with extruding near first/last point. Added the tests and updated with addressing other comments. I have also tested with various scenarios. Can you take a look?, Thank you.

sivako commented on docs/api-reference/mode-handlers/overview.md:
```diff
@@ -16,6 +16,12 @@ No edits are possible, but selection is still possible.
+
+* Mode name: `extrude`
+
+User can move edge. drag from any where between 2 points in edge.
```
I just updated.

sivako commented on modules/core/src/test/lib/mode-handlers/extrude-handler.test.js:
```diff
@@ -0,0 +1,227 @@
+// @flow
+/* eslint-env jest */
```
Agree, I have added the tests for handleClick, handleStartDragging, handlePointerMove as well.

sivako commented on docs/api-reference/layers/editable-geojson-layer.md:
```diff
@@ -139,6 +139,12 @@ The `onEdit` event is the core event provided by this layer and must be handled
 
   * `translated`: A feature finished translating (e.g. user finished dragging).
 
+  * `startExtruding`: A feature started extruding.
```
Ok, sure, I just updated. Thank you.

sivako commented:
Thank you @supersonicclay @georgios-uber. I have updated doc and tests. Can you please merge to master and create new tag for release?
